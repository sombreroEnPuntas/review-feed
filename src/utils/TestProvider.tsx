import React from 'react'
import { Provider } from 'react-redux'
import { ThemeWrapper } from 'retro-ui'

import store from '../data'

const TestProvider = ({ children }) => (
  <ThemeWrapper>
    <Provider store={store}>{children}</Provider>
  </ThemeWrapper>
)

export default TestProvider
