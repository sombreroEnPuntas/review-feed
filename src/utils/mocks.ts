import {
  ModelJWTResponse,
  ModelAPIError,
  ModelTheme,
  ModelReview,
} from '../client/api'

export const internalError = new Error(`Wut?. SyntaxError or NetworkError`)

export const errorAPIResponse: ModelAPIError = {
  code: 418,
  message: `The resulting entity body MAY be short and stout.`,
}

export const unauthorizedAPIResponse: ModelAPIError = {
  code: 401,
  message: `You shall not pass!`,
}

export interface MockData {
  login?: ModelJWTResponse
  themes?: ModelTheme[]
  reviews?: ModelReview[]
}
export class MockService {
  data
  error

  constructor(data: MockData, error: 'internal' | 'API' | 'EXPIRED_TOKEN') {
    this.data = data
    this.error = error
  }

  private genericErrorResponse = () => {
    if (this.error === 'API') {
      throw { json: () => Promise.resolve(errorAPIResponse) }
    } else if (this.error === 'EXPIRED_TOKEN') {
      throw { json: () => Promise.resolve(unauthorizedAPIResponse) }
    } else if (this.error === 'internal') {
      throw internalError
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  loginPost = async (_username: string, _password: string) => {
    this.genericErrorResponse()

    return this.data?.login
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  apiThemesGet = async (_limit: number, _offset: number) => {
    this.genericErrorResponse()

    return { data: this.data?.themes }
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  apiReviewsGet = async (_limit: number, _offset: number, _themeId: number) => {
    this.genericErrorResponse()

    return { data: this.data?.reviews }
  }
}

export const reviewMock = {
  id: 59435909,
  created_at: '2020-08-20T11:54:52Z',
  comment: 'Good, Bad, and Ugly.',
  themes: [
    { theme_id: 1000, sentiment: 1 },
    { theme_id: 1001, sentiment: 0 },
    { theme_id: 1002, sentiment: -1 },
  ],
}

export const themeListMock = [
  { id: 1000, name: 'General' },
  { id: 1001, name: 'Other' },
  { id: 1002, name: 'Unspecific' },
]
