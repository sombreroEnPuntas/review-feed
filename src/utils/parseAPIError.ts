import { ModelAPIError } from '../client/api'
import { clearToken } from '../data/login'

const UNAUTHORIZED_ERROR_CODE = 401

const parseAPIError = async (e: unknown, dispatch) => {
  let message: string

  // `instanceof Response` is not reliable when contexts are different :)
  // duct-taping is ok tho cuz we just need to unwrap a server response
  if ((e as Body).json) {
    const errorResponse: ModelAPIError = await (e as Body).json()
    if (errorResponse.code === UNAUTHORIZED_ERROR_CODE) {
      dispatch(clearToken())
    }
    message = `${errorResponse.code}: ${errorResponse.message}`
  } else {
    message = `${(e as Error).message}`
  }

  return message
}

export default parseAPIError
