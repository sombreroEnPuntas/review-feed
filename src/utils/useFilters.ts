import { useDispatch, useSelector } from 'react-redux'

import { ModelTheme } from '../client/api'
import { getThemeFilters, setThemeFilters } from '../data/filters'
import { getThemes } from '../data/themes'

const useFilters = () => {
  const themes: ModelTheme[] = useSelector(getThemes)
  const themeFilters: number[] = useSelector(getThemeFilters)

  const dispatch = useDispatch()

  const filter = (selection: number[]) => dispatch(setThemeFilters(selection))
  const reset = () => dispatch(setThemeFilters([]))

  return { filter, reset, themes, themeFilters }
}

export default useFilters
