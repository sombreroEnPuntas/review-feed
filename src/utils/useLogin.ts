import { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import {
  clearToken,
  getError,
  getToken,
  initialToken,
  setToken,
} from '../data/login'

const useLogin = () => {
  const dispatch = useDispatch()

  const error: string = useSelector(getError)
  const token: string = useSelector(getToken)

  useEffect(() => {
    dispatch(initialToken())
  }, [dispatch])

  const login = (username: string, password: string) => {
    dispatch(setToken(username, password))
  }

  const logout = () => {
    dispatch(clearToken())
  }

  return { error, login, logout, token }
}

export default useLogin
