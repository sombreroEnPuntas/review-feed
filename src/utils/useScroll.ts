import { useState, useEffect } from 'react'

const useScroll = (callback) => {
  const [loading, setLoading] = useState(false)
  const [scroll, setScroll] = useState(0)

  function handleScroll() {
    if (
      window.innerHeight + document.documentElement.scrollTop !==
        document.documentElement.offsetHeight ||
      loading
    )
      return

    setScroll(document.documentElement.scrollTop)
    setLoading(true)
  }

  useEffect(() => {
    window.addEventListener('scroll', handleScroll)

    return function cleanUp() {
      window.removeEventListener('scroll', handleScroll)
    }
  }, [])

  useEffect(() => {
    if (!loading) return
    ;(async () => {
      await callback()
      setLoading(false)
      document.documentElement.scrollTop = scroll
    })()
  }, [loading])
}

export default useScroll
