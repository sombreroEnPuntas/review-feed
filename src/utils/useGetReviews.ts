import { useRef, useEffect, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'

import { DefaultApi, ModelReview, ModelTheme } from '../client/api'
import { Configuration } from '../client/configuration'
import parseAPIError from './parseAPIError'
import { getToken } from '../data/login'

interface ModelThemeResponse {
  data: ModelTheme[]
}

interface ModelReviewResponse {
  data: ModelReview[]
}

const useGetReviews = () => {
  const service = useRef<DefaultApi>()

  const dispatch = useDispatch()
  const token: string = useSelector(getToken)

  const [error, setError] = useState<string>()
  const [loading, setLoading] = useState(false)
  const [reviews, setReviews] = useState<ModelReview[]>()
  const [themes, setThemes] = useState<ModelTheme[]>()

  const getThemes = async ({
    limit,
    offset,
  }: {
    limit?: number
    offset?: number
  }) => {
    setLoading(true)
    let message = null

    try {
      // wrong types on the model :(
      const response: unknown = await service.current.apiThemesGet(
        limit,
        offset
      )
      setThemes((response as ModelThemeResponse).data)
    } catch (e) {
      message = await parseAPIError(e, dispatch)
    }

    setError(message)
    setLoading(false)
  }

  const getReviews = async ({
    limit = 10,
    offset,
    themeId,
  }: {
    limit?: number
    offset?: number
    themeId?: number
  } = {}) => {
    setLoading(true)
    let message = null

    try {
      // wrong types on the model :(
      const response: unknown = await service.current.apiReviewsGet(
        limit,
        offset,
        themeId
      )
      setError(null)
      setReviews((prevData) => [
        ...(prevData || []),
        ...(response as ModelReviewResponse).data,
      ])
    } catch (e) {
      message = await parseAPIError(e, dispatch)
    }

    setError(message)
    setLoading(false)
  }

  // instantiate the service when new tokens come only
  useEffect(() => {
    ;(async () => {
      if (!token) return

      const config: Configuration = { apiKey: `Bearer ${token}` }
      service.current = new DefaultApi(config)

      if (!themes) {
        // initializes all, since it's just text
        await getThemes({ limit: 100 })
      }
      if (!reviews) {
        await getReviews()
      }
    })()

    return function cleanUp() {
      service.current = undefined
    }
  }, [token])

  return { error, loading, reviews, themeList: themes, getReviews }
}

export default useGetReviews
