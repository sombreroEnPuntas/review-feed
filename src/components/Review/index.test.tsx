import { render } from '@testing-library/react'
import React from 'react'
import { ThemeWrapper } from 'retro-ui'

import TestedComponent, { Props } from '.'

// Utils
import { reviewMock, themeListMock } from '../../utils/mocks'
// Mocks
const getProps = (customProps) => ({
  review: { ...reviewMock },
  themeList: [...themeListMock],
  ...customProps,
})

const stageTest = (customProps?: Props) => {
  return render(
    <ThemeWrapper>
      <TestedComponent {...getProps(customProps)} />
    </ThemeWrapper>
  )
}

describe('Review', () => {
  it(`renders a Review`, async () => {
    const { container, getByText } = stageTest()

    const review = getByText(/Good, Bad, and Ugly./)
    const sentiment1 = getByText(/General/)
    const sentiment2 = getByText(/Other/)
    const sentiment3 = getByText(/Unspecific/)

    expect(review).toBeInTheDocument()
    expect(sentiment1).toBeInTheDocument()
    expect(sentiment2).toBeInTheDocument()
    expect(sentiment3).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
