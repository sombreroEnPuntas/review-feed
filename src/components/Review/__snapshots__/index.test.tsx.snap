// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review renders a Review 1`] = `
.c1 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  display: block;
  margin: 0.5em;
  max-width: 450px;
  overflow-x: auto;
  padding: 0.5em;
  white-space: pre-wrap;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #212529;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  line-height: calc(2 * 1.5em);
}

.c0 {
  background-color: white;
  font-size: 0.5em;
  left: 2em;
  line-height: 1.5em;
  padding: 0 0.5em;
  position: relative;
  top: 1.2em;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
}

.c3 {
  display: block;
  width: 100%;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c5 {
  background-color: #c9d68f;
  margin: 0.5em;
  padding: 0 0.5em;
  white-space: nowrap;
}

.c6 {
  background-color: #E9C46A;
  margin: 0.5em;
  padding: 0 0.5em;
  white-space: nowrap;
}

.c7 {
  background-color: #ffd7cf;
  margin: 0.5em;
  padding: 0 0.5em;
  white-space: nowrap;
}

.c8 {
  margin: 0 1em;
}

<div>
  <article>
    <div
      class="c0"
    >
      8/20/2020
    </div>
    <pre
      class="c1"
      type="black"
    >
      <code
        class="c2"
        type="black"
      >
        <div
          class="c3"
        >
          <div
            class="c4"
          >
            <span
              class="c5"
            >
              General
            </span>
            <span
              class="c6"
            >
              Other
            </span>
            <span
              class="c7"
            >
              Unspecific
            </span>
          </div>
          <hr />
          <div
            class="c8"
          >
            Good, Bad, and Ugly.
          </div>
        </div>
      </code>
    </pre>
  </article>
</div>
`;
